plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/MethodEmbedding.kt

```diff
@@ -5,24 +5,110 @@
+import org.jetbrains.kotlin.fir.declarations.utils.isInline
+import org.jetbrains.kotlin.fir.expressions.FirExpression
+import org.jetbrains.kotlin.fir.symbols.SymbolInternals
+import org.jetbrains.kotlin.fir.symbols.impl.FirFunctionSymbol
+import org.jetbrains.kotlin.formver.conversion.*
+import org.jetbrains.kotlin.formver.domains.convertType
+import org.jetbrains.kotlin.formver.viper.MangledName
+import org.jetbrains.kotlin.formver.viper.ast.*
 
-class MethodEmbedding(
+interface MethodEmbedding : MethodSignatureEmbedding {
+    val shouldIncludeInProgram: Boolean
+    val viperMethod: Method
+    fun convertBody(ctx: ProgramConverter)
+    fun insertCall(argsFir: List<FirExpression>, ctx: StmtConversionContext<ResultTrackingContext>): Exp.LocalVar
+}
+
+abstract class BaseMethodEmbedding(
     val signature: MethodSignatureEmbedding,
-    val preconditions: List<Exp>,
-    val postconditions: List<Exp>,
-    val body: Stmt.Seqn?,
-    val isInline: Boolean,
-) : MethodSignatureEmbedding by signature {
-    val shouldIncludeInProgram = !isInline || body != null
-    val viperMethod = UserMethod(
-        name,
-        formalArgs.map { it.toLocalVarDecl() },
-        returnVar.toLocalVarDecl(),
-        preconditions,
-        postconditions,
-        body
-    )
+) : MethodEmbedding, MethodSignatureEmbedding by signature {
+    abstract val preconditions: List<Exp>
+    abstract val postconditions: List<Exp>
+    abstract val body: Stmt.Seqn?
+
+    override val viperMethod
+        get() = UserMethod(
+            name,
+            formalArgs.map { it.toLocalVarDecl() },
+            returnVar.toLocalVarDecl(),
+            preconditions,
+            postconditions,
+            body
+        )
+}
+
+class UserMethodEmbedding(
+    signature: MethodSignatureEmbedding,
+    override val preconditions: List<Exp>,
+    override val postconditions: List<Exp>,
+    val symbol: FirFunctionSymbol<*>,
+) : BaseMethodEmbedding(signature) {
+    override var body: Stmt.Seqn? = null
+    override val shouldIncludeInProgram
+        get() = !symbol.isInline || body != null
+
+    @OptIn(SymbolInternals::class)
+    override fun convertBody(ctx: ProgramConverter) {
+        val methodCtx = object : MethodConversionContext, ProgramConversionContext by ctx {
+            override val signature: MethodSignatureEmbedding = this@UserMethodEmbedding
+
+            // It seems like Viper will propagate the weakest precondition through the label correctly even in the absence of
+            // explicit invariants; we only need to add those if we want to make a stronger claim.
+            override val returnLabel: Label = Label(ReturnLabelName, listOf())
+            override val returnVar: VariableEmbedding = VariableEmbedding(ReturnVariableName, signature.returnType)
+
+            override val preconditions: List<Exp> = this@UserMethodEmbedding.preconditions
+            override val postconditions: List<Exp> = this@UserMethodEmbedding.postconditions
```

Comment:
Now that the pre- and post-conditions are part of the signature I don't think they should be abstract fields required by the `MethodConversionContext` interface.